drivers/spi/spi_mchp_mss.c

```diff
@@ -0,0 +1,496 @@
+	while (transfer_idx < count) {
+		if (!(mss_spi_read(cfg, MSS_SPI_REG_STATUS) & MSS_SPI_STATUS_RXFIFO_EMPTY)) {
+			rx_raw = mss_spi_read(cfg, MSS_SPI_REG_RX_DATA);
+			if (transfer_idx >= tr_len) {
+				if (spi_context_rx_buf_on(ctx)) {
+					UNALIGNED_PUT(rx_raw, (uint8_t *)ctx->rx_buf);
+					spi_context_update_rx(ctx, 1, 1);
+				}
+			}
+			++transfer_idx;
+		}
+
+		if (!(mss_spi_read(cfg, MSS_SPI_REG_STATUS) & MSS_SPI_STATUS_TXFIFO_FULL)) {
+			if (spi_context_tx_buf_on(ctx)) {
+				data8 = ctx->tx_buf[0];
+				mss_spi_write(cfg, MSS_SPI_REG_TX_DATA, data8);
+				spi_context_update_tx(ctx, 1, 1);
+			} else {
+				mss_spi_write(cfg, MSS_SPI_REG_TX_DATA, 0x0);
+			}
+		}
+	}
+}
+
+static inline int mss_spi_select_slave(const struct mss_spi_config *cfg, int cs)
+{
+	uint32_t slave;
+	uint32_t reg = mss_spi_read(cfg, MSS_SPI_REG_SS);
+
+	slave = (cs >= MSS_SPI_MIN_SLAVE && cs <= MSS_SPI_MAX_SLAVE) ? (1 << cs) : 0;
+	reg &= ~MSS_SPI_SSEL_MASK;
+	reg |= slave;
+
+	mss_spi_write(cfg, MSS_SPI_REG_SS, reg);
+
+	return 0;
+}
+
+static inline void mss_spi_activate_cs(struct mss_spi_config *cfg)
+{
+	uint32_t reg = mss_spi_read(cfg, MSS_SPI_REG_SS);
+
+	reg |= MSS_SPI_SSELOUT;
+	mss_spi_write(cfg, MSS_SPI_REG_SS, reg);
+}
+
+static inline void mss_spi_deactivate_cs(const struct mss_spi_config *cfg)
+{
+	uint32_t reg = mss_spi_read(cfg, MSS_SPI_REG_SS);
+
+	reg &= ~MSS_SPI_SSELOUT;
+	mss_spi_write(cfg, MSS_SPI_REG_SS, reg);
+}
+
+static inline int mss_spi_clk_gen_set(const struct mss_spi_config *cfg,
+				      const struct spi_config *spi_cfg)
+{
+	uint32_t idx, clkrate, val = 0, speed;
+
+	if (spi_cfg->frequency > cfg->clock_freq) {
+		speed = cfg->clock_freq / 2;
+	}
+
+	for (idx = 1; idx < 16; idx++) {
+		clkrate = cfg->clock_freq / (2 * idx);
+		if (clkrate <= spi_cfg->frequency) {
+			val = idx;
+			break;
+		}
+	}
+
+	mss_spi_write(cfg, MSS_SPI_REG_CLK_GEN, val);
+
+	return 0;
+}
+
+static inline int mss_spi_hw_mode_set(const struct mss_spi_config *cfg, unsigned int mode)
+{
+	uint32_t control = mss_spi_read(cfg, MSS_SPI_REG_CONTROL);
+
+	/* set the mode */
+	if (mode & SPI_MODE_CPHA) {
+		control |= MSS_SPI_CONTROL_SPH;
+	} else {
+		control &= ~MSS_SPI_CONTROL_SPH;
+	}
+
+	if (mode & SPI_MODE_CPOL) {
+		control |= MSS_SPI_CONTROL_SPO;
+	} else {
+		control &= ~MSS_SPI_CONTROL_SPO;
+	}
+
+	mss_spi_write(cfg, MSS_SPI_REG_CONTROL, control);
+
+	return 0;
+}
+
+static void mss_spi_hw_release(const struct mss_spi_config *cfg)
+{
+	uint32_t control = mss_spi_read(cfg, MSS_SPI_REG_CONTROL);
+
+	control &= ~MSS_SPI_CONTROL_ENABLE;
+	mss_spi_write(cfg, MSS_SPI_REG_CONTROL, control);
+}
+
+static void mss_spi_interrupt(const struct device *dev)
+{
+	const struct mss_spi_config *cfg = dev->config;
+	struct mss_spi_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int intfield = mss_spi_read(cfg, MSS_SPI_REG_MIS) & 0xf;
+
+	if (intfield == 0) {
+		return;
+	}
+
+	/* clear the interrupt conditions */
```

Comment:
Is there a reason why you specifically handle each interrupt the same way but separately? 

Couldn't it be handled this way:
```
mss_spi_write(cfg, MSS_SPI_REG_INT_CLEAR, intfield);
spi_context_complete(ctx, dev, 0);
```

And more importantly: where do you actually act on the relevant interrupt?
Why the logic of mss_spi_readwr_fifo() is not being called in ISR mode?

If this controller is able to run in interrupt mode, logically you would configure/enable/unmask INTs and... in the interrupt handler here, you would move bytes in/out depending on the interrupt type.

Reply:
> Is there a reason why you specifically handle each interrupt the same way but separately?
> 
> Couldn't it be handled this way:
Thanks Tomasz for your time to review this PR.
Yes the below way is the cleanest way. I will update it.
> 
> ```
> mss_spi_write(cfg, MSS_SPI_REG_INT_CLEAR, intfield);
> spi_context_complete(ctx, dev, 0);
> ```
> 
> And more importantly: where do you actually act on the relevant interrupt? Why the logic of mss_spi_readwr_fifo() is not being called in ISR mode?
> 
> If this controller is able to run in interrupt mode, logically you would configure/enable/unmask INTs and... in the interrupt handler here, you would move bytes in/out depending on the interrupt type.
I agree with your explanation about generically handling tx and rx from interrupt handler.
But we configure the controller in such a way that, the TX DONE will be set once the configured number of transmit bytes are sent and the RXRDY will be set once all the configured number of rx bytes are read from the RXFIFO(from the mss_spi_readwr_fifo() functions). hence in the interrupt handler we just acknowledge the bits as the entire data transmission and reception is handled inside mss_spi_readwr_fifo().
But agree that, in this case the interrupt handler is very small as mentioned by you like mss_spi_write(cfg, MSS_SPI_REG_INT_CLEAR, intfield);

Reply:
so you loose the interest of having the controller running on interrupt mode, no? It's like polling mode